ggu/main.c

```diff
@@ -16,11 +16,295 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <stdbool.h>
+
+#include <rte_ip.h>
+#include <rte_udp.h>
+#include <rte_log.h>
+#include <rte_lcore.h>
+#include <rte_malloc.h>
+#include <rte_ethdev.h>
+#include <rte_atomic.h>
+
+#include "gatekeeper_gk.h"
 #include "gatekeeper_ggu.h"
+#include "gatekeeper_net.h"
+#include "gatekeeper_main.h"
+#include "gatekeeper_config.h"
+
+static rte_atomic16_t num_ggu_conf_alloc = RTE_ATOMIC16_INIT(0);
+
+static int
+policy_parse_next(uint16_t *proto_ptr, uint8_t *policy_ptr,
+	struct ggu_policy *policy)
+{
+	if (proto_ptr == NULL || policy_ptr == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"ggu: invalid pointor when parsing policies at %s\n", __func__);
+		return -1;
+	}
+
+	policy->flow.proto = rte_be_to_cpu_16(*(uint16_t *)proto_ptr);
+	if (policy->flow.proto == ETHER_TYPE_IPv4)
+		rte_memcpy(&policy->flow.f.v4, policy_ptr,
+			sizeof(policy->flow.f.v4));
+	else if (policy->flow.proto == ETHER_TYPE_IPv6)
+		rte_memcpy(&policy->flow.f.v6, policy_ptr,
+			sizeof(policy->flow.f.v6));
+	else {
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"ggu: parse policy with unknown network layer protocol %hhu\n",
+			policy->flow.proto);
+		return -1;
+	}
+
+	rte_memcpy(&policy->params, policy_ptr, sizeof(policy->params));
+
+	return 0;
+}
+
+static void
+process_single_policy(struct ggu_policy *policy, struct ggu_config *ggu_conf)
+{
+	/*
+	 * Identify which GK block is responsible for the
+	 * pair <Src, Dst> in the decision.
+	 */
+	int block_id = get_responsible_block(&policy->flow, ggu_conf->gk);
+	struct gk_cmd_entry entry;
+
+	if (block_id < 0)
+		return;
+	/*
+	 * Obtain mailbox of that GK block,
+	 * and send the policy decision to the GK block.
+	 */
+	entry.op = GGU_POLICY_ADD;
+	rte_memcpy(&entry.u.ggu.flow, &policy->flow, sizeof(policy->flow));
+	entry.u.ggu.params.state = policy->params.state;
+	entry.u.ggu.params.tx_rate_kb_sec =
+		rte_be_to_cpu_32(policy->params.tx_rate_kb_sec);
+	entry.u.ggu.params.cap_expire_sec =
+		rte_be_to_cpu_32(policy->params.cap_expire_sec);
+	entry.u.ggu.params.renewal_step_ms =
+		rte_be_to_cpu_32(policy->params.renewal_step_ms);
+
+	send_gk_cmd(block_id, &entry, ggu_conf->gk);
+}
+
+static void
+process_single_packet(struct rte_mbuf *pkt, struct ggu_config *ggu_conf)
+{
+	uint8_t j;
+	uint8_t *data;
+	uint8_t *policy_ptr;
+	uint8_t num_policies;
+	uint16_t *proto_ptr;
+	uint16_t ether_type;
+	struct ether_hdr *eth_hdr;
+	struct ipv4_hdr *ip4hdr;
+	struct ipv6_hdr *ip6hdr;
+	struct udp_hdr *udphdr;
+
+	eth_hdr = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	ether_type = rte_be_to_cpu_16(eth_hdr->ether_type);
+
+	switch (ether_type) {
+	case ETHER_TYPE_IPv4:
+		ip4hdr = rte_pktmbuf_mtod_offset(pkt, 
+			struct ipv4_hdr *, sizeof(struct ether_hdr));
+		if (ip4hdr->next_proto_id != IPPROTO_UDP) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"ggu: received non-UDP packets, ntuple filter bug!\n");
```

Comment:
Please edit the string to be the one below:

`"ggu: received non-UDP packets, IPv4 ntuple filter bug!\n"`

This would help a lot debugging if this bug is ever encountered.